455217 [C#テキスト]1章演習問題　作り直し

kurosu-being opened the pull request:
Gitでトラブルが起きたためブランチを作り直しました。

第1章 オブジェクト指向プログラミングの基礎

オブジェクト指向プログラミングの基本的な概念
１．クラスと構造体
２.　値型と参照型の違い
３．静的メンバーと静的クラス
４．名前空間と継承
などを学習しました。

kurosu-being committed:
章末問題第1章1.1-1～1.2-1までを実装しました

kurosu-being committed:
演習問題第1章1.2-2～1.3-4までを実装しました。

kurosu-being committed:
#455217 [C#テキスト]1章演習問題

メンバー変数などの名付けを一部変更しました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題

全てのcsにおいて、不要なusingを消去し、コードを整形しました。
全てのPrgram.csをそれぞれのファイル名に対応した名前に変更しました。（例：Program.cs → Program1.1-1,1.1-2.1.1-3.cs）また、その際不要な章末問題1.1-3ファイルを削除し、第1章1.1の問題をすべて同じファイルに統合しました。
Practice1.1,1.1-2,1.1-3において、クラス、プロパティにそれぞれにヘッダーコメントを付けました。
Practice1.1,1.1-2,1.1-3において、Console.WriteLineに"円です"が表示されるように追加し、結果の単位が表示されるように修正しました。
Practice1.1,1.1-2,1.1-3において、ローカル変数の命名をdorayaki→wdorayakiに変更しました。また、以降問題についてもmyClass,mystudentを→wmyClass,wstudentに変更しました。

kurosu-being committed:
#455217 [C#テキスト]1章演習問題

全てのProgram.csにおいて、問題の表記方法を変更しました。（例　Program1.2-1,1.2-2→Program1-2-1,1-2-2）
Program1-1-1,1-1-2,1-1-3において、usingの並びを変更しました。
Program1-2-1,1-2-2において、問題文の修正を行いました（方→型）
Program1-2-1,1-2-2において、メソッドの間に空行を追加しました。また、空行の削除を行いました。
問３においてStudent.csの
"{0}は、{1}年{2}組に所属しています",student.Name, student.Grade, student.ClassNumber
を文字列補間式で書き換えました。

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Program1-1-1,1-1-2,1-1-3.cs:
```diff
@@ -0,0 +1,15 @@
+        //Practice1.1-2 どら焼きオブジェクトの消費税額を求め、コンソールに出力するコードを書いてください
+        //Practice1.1-3 Productクラスが属する名前空間を別の名前空間に変更し、Mainメソッドから呼び出すようにしてください。ただし、MainメソッドのあるProgramクラスの名前空間はそのままとしてください。
+        static void Main(string[] args) {
+            Product wdorayaki = new Product(98, "どら焼き", 210);
+            Console.WriteLine("消費税額は" + wdorayaki.GetTax() + "円です");
```
接頭辞w の後は、大文字になるので、`wDorayaki` が正しいですね。
他の箇所も同様に、修正をお願いします。

kurosu-being replied:
ありがとうございます！コーディング規約の例をよく見ていませんでした。
修正を行いました。

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Product.cs:
```diff
@@ -0,0 +1,34 @@
+        /// </summary>
+        public int Price { get; set; }
+        /// <summary>
+        /// コンストラクタ
+        /// </summary>
+        /// <param name="code"></param>
+        /// <param name="name"></param>
+        /// <param name="price"></param>
+        public Product(int code, string name, int price) {
+            this.Code = code;
```
ヘッダーコメントの追加ありがとうございます！
`<param name="">`は、引数を表しています。
引数がそれぞれ何を指しているのか、プロパティと同様に説明を書いてみてください。

また、コーディング規約に則って、引数にも接頭辞をつけましょう！

kurosu-being replied:
ご指摘ありがとうございます！
コーディング規約に則って、引数に接頭辞を追加しました。
引数の説明はプロパティと同じに変更しましたが、”商品名の引数”など名前を変える方が良いでしょうか？

k-miyahara-being replied:
> 引数の説明はプロパティと同じに変更しましたが、”商品名の引数”など名前を変える方が良いでしょうか？
これはすでに解決済みですかね？

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Product.cs:
```diff
@@ -0,0 +1,34 @@
+            this.Name = name;
+            this.Price = price;
+        }
+        /// <summary>
+        /// 消費税を求める（消費税率は10%）
+        /// /// </summary>
+        /// <returns>int</returns>
```
`<returns>`は、戻り値を表しています。
ここには、戻り値が何を指しているかの説明を書いてください。

kurosu-being replied:
ご指摘ありがとうございます！戻り値の説明を追加しました。

k-miyahara-being commented on 章末問題第1章/Practice1.3-3, 1.3-4/Student.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+
+namespace Practice1._3_1 {
+    class Student : Person {
+        public int Grade { get; set; }
+        public int ClassNumber { get; set; }
+        public Student() {
+            this.Name = "黒子";
+            this.Birthday = new DateTime(2000, 2, 2);
+            this.Grade = 3;
+            this.ClassNumber = 10;
+        }
```
ヘッダーコメントについての修正漏れがあります！
指摘があった場合は、指摘箇所だけでなく他にも同様の箇所がないか見てみましょう。

kurosu-being replied:
ご指摘ありがとうございます。全てのcsファイルを確認し、publicメソッド、プロパティを中心にヘッダーコメントを記入しました。

k-miyahara-being commented on 章末問題第1章/Practice1.3-3, 1.3-4/Person.cs:
```diff
@@ -0,0 +1,15 @@
+        public string Name { get; set; }
+        public DateTime Birthday { get; set; }
+        public int GetAge() {
+            DateTime today = DateTime.Today;
+            int age = today.Year - Birthday.Year;
+            if (today < Birthday.AddYears(age))
+                age--;
+            return age;
```
コーディング規約を適用しましょう！

また、if文は1行にまとめてしまったほうが見やすい時があります。
今回のように、分岐した中身が`age--;` しか処理がないので、↓のように1行にまとめておきましょう。
`if (today < Birthday.AddYears(age)) age--;`

kurosu-being replied:
ご指摘ありがとうございます。
問3のPerson.csにおいて、コーディング規約に則りローカル変数の名前を
today →wToday, age → wAge に変更しました。
また、ifを一行に纏めました。

k-miyahara-being commented on 章末問題第1章/Practice1.3-1/Program1-3-1,1-3-2.cs:
```diff
@@ -0,0 +1,13 @@
+        //Practice1.3-3 2で生成したインスタンスの各プロパティの値をコンソールに出力するコードを書いてください。
+        static void Main(string[] args) {
+            Student wtudent = new Student();
+            Console.WriteLine("{0}は、{1}年{2}組に所属しています", wtudent.Name, wtudent.Grade, wtudent.ClassNumber);
```
https://github.com/kurosu-being/TextC/pull/1#discussion_r1358123966

kurosu-being replied:
ご指摘ありがとうございます。
文字列補間式の変更がPractice1-3-3,1-3-4にしか反映しておりませんでした。
全てのcsで変更し、反映しました。

k-miyahara-being commented on 章末問題第1章/Practice1.3-1/Student.cs:
```diff
@@ -0,0 +1,14 @@
+    class Student : Person {
+        public int Grade { get; set; }
+        public int ClassNumber { get; set; }
+        public Student(){
+            this.Name = "黒子";
+            this.Birthday = new DateTime(2000, 2, 2);
+            this.Grade = 3;
+            this.ClassNumber = 10;
+        }
```
このままだと、Studentクラスのインスタンスはすべて中身が同じになってしまいます。
引数で受け取って、各プロパティに代入する形になっているとよいですね。

kurosu-being replied:
ご指摘ありがとうございます！
引数で受け取って、各プロパティに代入する形に変更いたしました。

k-miyahara-being reviewed (changes requested):
コード整形のおかげで、かなり見やすくなりました！
もう少し修正をお願いします

kurosu-being committed:
コーディング規約に則りローカル変数の名前を変更しました。
指摘部分のコントラクタの引数の説明を追加しました。

kurosu-being committed:
ヘッダーコメント（MyClass.cs)を追加しました。
returnについてのヘッダーコメントを記入しました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題　作り直し

Program1-3-,1-3-2.cs →Program1-3-,1-3-2.3-3.cs　に名前を変更しました。（1-3-3が抜けておりました）
問3全てのPerson.csにおいて、コーディング規約に則りローカル変数の名前を命名しなおしました。（today →wToday, age → wAge)
問3全てのPerson.csにおいて、ifの列数を整理しました。
問3全てのProgram.csにおいて文字列補間式でコンソールに表示されるように変更しました。
問3全てのStudent.csにおいてインスタンスの中身が同じになる問題を解消し、引数で受け取って、各プロパティに代入する形に変更しました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題　作り直し

問1のProgram.csにおいてusingの並びを変更しました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題　作り直し

問3のProgram.csにおいて、ローカル変数の名前を変更しました　（name →wName, grade →wGrade, classnumber →wClassnumber)

kurosu-being committed:
455217 [C#テキスト]1章演習問題　作り直し

問1のProduct.cs1に商品の消費税の小数点以下を四捨五入するするように実装しました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題　作り直し

問1のProduct.csにおいて、プロパティの商品コードのprivate set;を削除しました。

m-hotta-being commented on 章末問題第1章/Practice1.2-1/MyClass.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace Practice1._2_1 {
```
【必要なら修正してください】
名前空間 (namespace) は以下のように `.` で区切ることで階層 `System` - `Collections` - `Generic` を表現できます。
```
System.Collections.Generic
```

本題ですが `Practice1._2_1` は `Practice1` - `_2_1` なのでしょうか？
`_2_1` が `_` で始まっていてあまり見ない形なので気になりました。

kurosu-being replied:
ご指摘ありがとうございます！
意図したものではなく、namespaceを変更すると
｛が必要です
というエラーが出てきてしまうため放置しておりました。
原因を調べてみようと思います。

m-hotta-being replied:
> 原因を調べてみようと思います。

きっと、このページが参考になると思います。

* [C# 識別子の名前付け規則と表記規則](https://learn.microsoft.com/ja-jp/dotnet/csharp/fundamentals/coding-style/identifier-names)
  * [名前指定の規則](https://learn.microsoft.com/ja-jp/dotnet/csharp/fundamentals/coding-style/identifier-names#naming-rules)

m-hotta-being commented on 章末問題第1章/Practice1.2-1/Program1-2-1,1-2-2.cs:
```diff
@@ -0,0 +1,25 @@
+        /// 2つのオブジェクトの内容（プロパティの値)をコンソールに表示するメソッド
+        /// </summary>
+        /// クラスの引数
+        /// <param name="vMC"></param>
```
惜しい！
`<param name="vMC">（ここに説明を書く）</param>` 

この形式のコメントの書き方は、このページが参考になると思います。

* [コード要素の XML コメントを挿入するには](https://learn.microsoft.com/ja-jp/visualstudio/ide/reference/generate-xml-documentation-comments?view=vs-2022)

kurosu-being replied:
ご指摘ありがとうございます。
修正いたしました！
頂いたURLも参考にコメントの書き方を勉強します！

m-hotta-being commented on 章末問題第1章/Practice1.3-3, 1.3-4/Person.cs:
```diff
@@ -0,0 +1,25 @@
+        public int GetAge() {
+            DateTime wToday = DateTime.Today;
+            int wAge = wToday.Year - Birthday.Year;
+            if (wToday < Birthday.AddYears(wAge)) wAge--;
```
【修正不要。ただのコメントです。】
`--wAge` という書き方があるのは知っていますか？
もしまだ知らないなら、どう違うか一度調べてみてください。

kurosu-being replied:
左右どちらに書くかで意味が違うのですね。
勉強になりました、ありがとうございます！

m-hotta-being reviewed (approved):
ちょっと見てみました。
確認してみてください。

kurosu-being committed:
455217 [C#テキスト]1章演習問題　作り直し

ヘッダーコメントで、引数の説明の位置を修正しました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題　作り直し

ヘッダーコメントの引数の説明で「○○の引数」→「○○」に変更漏れがあったところを修正しました

kurosu-being committed:
#455217 [C#テキスト]1章演習問題　

新しいプロジェクトに第1章の演習問題のコードを書き写しました。

k-miyahara-being commented on 章末問題第1章/演習問題第1章 remake/演習問題第1章問１/Product.cs:
```diff
@@ -0,0 +1,36 @@
+        /// <summary>
+        /// 消費税を求める（消費税率は10%）
+        /// /// </summary>
+        /// 商品価格の消費税をを消費税10%で計算した値
+        /// <returns>int</returns>
```
戻り値の説明も、他と同様にreturnsタグの間に書くようにしましょう。
`<returns>商品価格の消費税をを消費税10%で計算した値</returns>`
ほかの箇所にもあるので、確認お願いします

XMLについて：https://wa3.i-3-i.info/word1577.html

kurosu-being replied:
ご指摘ありがとうございます！
修正を行いました。戻り値の説明も文言で説明した方が良いのですね。

k-miyahara-being replied:
すみません、↑の例ではそのまま書いてしまいましたが、
メソッドの説明と被っている部分は冗長なのでなくてよいです。

↓みたいな感じでできるだけシンプルにしましょう。
```
/// <summary>
/// 消費税を求める（消費税率は10%）
/// </summary>
/// <returns>消費税額</returns>
```

kurosu-being replied:
ありがとうございます！シンプルな文言に書き換えます。

k-miyahara-being commented on 章末問題第1章/演習問題第1章 remake/演習問題第1章問3-4/Program1-3-4.cs:
```diff
@@ -0,0 +1,17 @@
+        //Practice1.3-4 2で生成したインスタンスをPerson型及びobject型の変数に代入できることを確認してください
+        static void Main(string[] args) {
+            Student wStudent = new Student("黒子", new DateTime(2000, 2, 2), 3, 10);
+            var vName = wStudent.Name;
+            var vGrade = wStudent.Grade;
+            var vClassnumber = wStudent.ClassNumber;
+            Console.WriteLine($"{vName}は、{vGrade}年{vClassnumber}組に所属しています");
```
ここは、わざわざ変数に代入する必要はないかなと思います。（コーディング規約違反でもありますね！）
`wStudent.Name`や`wStudent.Grade`は1か所でしか使用されていないからです。
メソッド内の複数箇所で使用するのであれば、書いてもらったように変数に代入して使いまわすとよいです。

kurosu-being replied:
ご指摘ありがとうございます！
変数を直接文字列補間式に書き込む方法に変更いたしました。
一か所でしか使用しない変数について考えが及んでいませんでした。勉強になります。

k-miyahara-being replied:
コメントアウトしたのは、コーディング規約の修正をしたことを伝えるためですかね？
そうでなければ、不要なコードは消しておきましょう！

kurosu-being replied:
修正を伝えるつもりでしたが、ここでも伝えているので不要ですね。
消しておきます！

k-miyahara-being reviewed (commented):
あと少しです！
修正をお願いします。

kurosu-being committed:
#455217 [C#テキスト]1章演習問題

問3において、文字列補間式の中身を直接変数値で記入しました。

kurosu-being committed:
#455217 [C#テキスト]1章演習問題

全csにおいて、戻り値の説明の位置をreturnタグの間に変更しました。

kurosu-being committed:
#455217 [C#テキスト]1章演習問題

すべてのcsで〈return〉の説明をシンプルにしました。
問3において、不要なコメント（以前の修正）を削除しました。

k-miyahara-being reviewed (approved):
OKです！

m-hotta-being reviewed (approved):
修正お疲れさまでした！
